File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> list = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(list, new ArrayList<>(), candidates, target, 0);
        return list;
    }
    
    private void backtrack(List<List<Integer>> list, List<Integer> temp_list, int[] nums, int remain, int start){
        if(remain < 0) return;
        else if(remain == 0) list.add(new ArrayList<>(temp_list));
        else {
            for(int i = start; i < nums.length; i++){
                temp_list.add(nums[i]);
                backtrack(list, temp_list, nums, remain - nums[i], i);
                temp_list.remove(temp_list.size()-1);
            }
        }
    }
    
}

File: 0039-combination-sum/0039-combination-sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        IList<IList<int>> res = new List<IList<int>>();
        Backtrack(res, new List<int>(), candidates, target, 0);
        return res;
    }

    public void Backtrack(IList<IList<int>> res, IList<int> temp, int[] candidates, int remain, int index)
    {
        if(remain < 0) return;
        if(remain == 0)
        {
            res.Add(new List<int>(temp));
        }
        else
        {
            for(int i = index; i < candidates.Length; i++)
            {
                temp.Add(candidates[i]);
                Backtrack(res, temp, candidates, remain - candidates[i], i);
                temp.RemoveAt(temp.Count - 1);
            }
        }
    }
}